### DISCUSSION TITLE
프로젝트 레이아웃

### DISCUSSION THREAD

**Discussion by psh2849** [euphony-io/check-in-speaker#5] [Ideas]
@euphony-io/euphoria 
먼저, 간략하게 레이아웃 틀만 만들어 보았습니다! 수정이나 보완이 필요한 점은 말씀해주세요!

**1. 앱을 받고 처음 실행할 때**
![1](https://user-images.githubusercontent.com/42669772/134333382-9cab1922-1c53-4af1-ac2e-8a6fd62543d3.PNG)

alertDialog를 통해 팝업 창을 띄워서 사용자의 전화번호를 입력 받으면 좋을 것 같습니다.
여기서 SharedPreference를 이용해서 안심 번호를 저장시키는 게 편할 것 같아요!

**2. 초기 화면**
![2](https://user-images.githubusercontent.com/42669772/134333607-d459acf6-7d80-453f-b56b-8b33478a3561.PNG)

여기서 팝업창을 닫고 난 후 사용자의 현재 위치를 받아서 텍스트뷰에 입력해주고,
SharedPreference에 저장되어 있는 안심 번호를 불러와서 텍스트뷰에 입력해주면 좋을 것 같다고 생각했습니다!

**3. 전송 화면**
![3](https://user-images.githubusercontent.com/42669772/134333807-606baf91-e7d7-4e77-b46a-dc59aab6a374.PNG)

위의 스피커 아이콘을 눌러서 전송 / 취소를 관리하면 좋을 것 같습니다.

안심 번호를 어떻게 만들지에 대해서 생각해 봤는데 안심번호 중복 부분을 어떻게 해결해야 할지 잘 모르겠습니다.... 안심 번호를 다시 전화 번호로 복호화하는 방법도 생각해봐야 할 것 같구요..
여러 의견 주시면 감사하겠습니다!

**Comment by jopopcorn** [comment]
연휴 보내느라 아직 작업 시작을 못했는데 레이아웃 참고하기 좋겠네요 감사합니다!!
스피커 아이콘도 괜찮은데 전 연령대가 사용하다 보니 좀 더 직관적인 UI가 필요할 것 같습니다.
`체크인 시작/중단`과 같은 버튼을 만들어서 음파가 생성되고 중단되게끔 하면 더 좋을 것 같아요!

기존에 사용하고 있는 개인안심번호는 QR 체크인을 제공하는 앱에서만 발급이 가능하더라고요.
이걸 그대로 갖다 쓰면 좋겠지만 공식적으로 세 군데에서만 승인해주어서 사용하기 어렵지 않을까 싶네요 ㅜ
그리고 개인안심번호 개발자의 인터뷰를 찾아보았는데, 7명에서 두 달 정도 소요했다는 걸 보니 단기간에 안심번호를 개발하기에는 어려울 듯 합니다 ㅜㅜ
완벽하게는 아니더라도 최대한 고민해보고 비슷하게 만들어봐도 좋을 것 같아요! 생각해보신 안심번호 로직 공유 부탁드려요!

**Reply by psh2849** [reply]
의견 감사합니다! 
말씀해주신 버튼 형식으로 만드는 것도 좋을 것 같습니다!

안심번호 로직은 저도 아직 생각 중이긴 한데..  기존의 사용자 정보들을 파이어베이스에 전화번호, 안심번호를 저장시켜서 관리하고, 
신규 사용자들에게 안심번호를 부여할 때는 숫자 4개 + 영문자 2개 정도로 이루어진 값을 랜덤 생성해서 생성된 안심번호와 파이어베이스에 저장되어있던 안심번호들을 비교해서 중복된 값이 있는지 없는지 검증해서 안심번호를 부여하는 방법을 생각해보았습니다..

이 방법이 간단하기는 할 것 같은데 파이어베이스에서 중복 체크하는 코드의 시간복잡도가 어느정도 일지 모르겠는데 데이터가 커지면 커질 수록 좀 느려질까봐 걱정이네요..

**Reply by jopopcorn** [reply]
로직은 간단하네요! 하지만 데이터가 많을 때 속도가 느려질 수도 있고, 랜덤으로 발급한 안심번호가 여러 번 중복되는 최악의 경우도 있을 것 같아요!
저도 다른 좋은 방법이 있는지 더 생각해보겠습니다!!

**Comment by jopopcorn** [comment]
<img width="1121" alt="check-in-speaker-design" src="https://user-images.githubusercontent.com/35393459/134382934-8202a366-c2d2-4cb0-aafd-45024c38ea2b.png">

성호님이 해주신 디자인과 QR 체크인 앱들 참고해서 UI 디자인 해보았습니다!
큐알이 없어서 허전한데 이건 어떻게 배치하는게 좋을까요.. 확장성을 위해 그대로 두는 게 좋을까요?
문구나 다른 디테일한 부분도 피드백 해주시면 감사하겠습니다~!

**Reply by psh2849** [reply]
깔끔하고 괜찮네요!
좀 비어있는 공간은 일단 그대로 둘까요..? 조금 허전해도 괜찮은거 같기도하고.. 
그리고 저희 Euphony가 볼륨을 키운 상태여야 통신이 가능하니까 그 점도 써두는게 좋지 않을까요?

**Reply by jopopcorn** [reply]
아하 볼륨을 키워달라는 문구 같은걸 추가하면 좋겠다는 말씀이시죠? 좋습니다!
공간이 너무 비어있으니 볼륨 조절하는 SeekBar를 추가해도 괜찮을 것 같은데 요건 어떻게 생각하시나요?

**Reply by psh2849** [reply]
볼륨 조절은 근데 디바이스에 버튼이 있어서 따로 안넣어줘도 괜찮지 않을까요??
저 공간이 너무 허전하시면 신호를 보내고 있다는 걸 직관적으로 표현해주는 아이콘 같은걸 넣는 것도 괜찮을 것 같고
아니면 아예 체크인 시작 버튼을 개인안심번호 텍스트뷰 바로 아래 붙이고 다 하나로 모아서 화면의 가운데에 위치하게끔 하는 것도 좋을 것 같습니다.

**Reply by jopopcorn** [reply]
네 좋습니다 그럼 아래와 같은 위젯을 추가로 해서 디자인 수정하겠습니다!
- `'음성 데이터로 체크인하기 위해 볼륨을 크게 키워주세요'`와 같은 안내 문구
- 체크인 시작 버튼을 눌렀을 때 신호를 보내는 중임을 표시하는 아이콘이나 프로그래스바 (빙글빙글 도는)
- 빈 공간이 많으므로 가운데에 배치하기 (뭐가 더 나은지 비교해서 결정하기)

**Reply by psh2849** [reply]
넵 감사합니다!

**Comment by jopopcorn** [comment]
<img width="1095" alt="speaker_기본_배치" src="https://user-images.githubusercontent.com/35393459/134632792-7758635a-5dc0-420c-a25c-0143863bf0c8.png">
<img width="1109" alt="speaker_가운데_배치" src="https://user-images.githubusercontent.com/35393459/134632812-27563c1f-7a7f-4871-a399-80efee2ee4a4.png">

@euphony-io/euphoria 
위에서 언급한 수정할 부분 반영해보았습니다! 위 아래 중 어떤 게 더 나은가요?!
프로그레스바는 임의로 아무 아이콘이나 넣은 것이고 실제로는 안드로이드 기본 아이콘 쓰면 될 것 같아요!

**Reply by psh2849** [reply]
두개 다 괜찮은데요? 보기에 편한건 위에거가 조금 더 나은 거 같네요!

**Reply by HACore** [reply]
굉장히 깔끔하게 잘 디자인 하셨네요👍
저는 1번이 더 보기 좋은것같습니다!

**Reply by kmsaidev** [reply]
저도 위에거가 좀 더 좋은 것 같습니다!!ㅎㅎ
